feat: add JSON-RPC client transport implementation

joshuafuller opened the pull request:
## Overview
Adds JSON-RPC 2.0 client transport support to the Go SDK. Implementation follows the structure and code style discussed in #73.

## Implementation
- All 10 required A2A JSON-RPC methods (message/send, message/stream, tasks/*, etc.)
- SSE streaming for real-time updates
- Thread-safe agent card caching
- 5-second default timeout (matches Python SDK)
- No external dependencies beyond stdlib
- Located in `a2aclient/jsonrpc.go` following the `grpcTransport` pattern

## Testing & Validation
Following the guidance in #73, I validated the implementation against the Python SDK reference implementation to ensure behavioral consistency:
- Compared method signatures and behavior with the [a2a-python transport](https://github.com/google/a2a-python/blob/main/src/a2a/client/transports/jsonrpc.py)
- Verified JSON-RPC request/response structures match spec
- Confirmed proper SSE parsing and error handling

Test suite includes 58.6% coverage with race detection:
```bash
$ go test -race ./a2aclient -run TestJSONRPC
PASS
ok      github.com/a2aproject/a2a-go/a2aclient  1.163s
```

All tests validate spec compliance including required field validation (Task: id/contextId/status.state, Message: id/role/parts).

## Files
- `a2aclient/jsonrpc.go` (558 lines) - Transport implementation
- `a2aclient/jsonrpc_test.go` (755 lines) - Test suite

Fixes #73

joshuafuller commented:
Thanks for the review feedback! I've addressed all the points in commit fb75ea1:

**High Priority:**
- Fixed concurrent test to verify exactly 1 network call (changed from lenient `> 3` check)

**Medium Priority:**
- Extracted all hardcoded strings to constants (JSON-RPC methods, headers, protocol version)
- Removed unnecessary context cancellation goroutines (HTTP client handles this)
- Simplified agent card locking to clean double-checked pattern

All tests pass with race detection. Net result: cleaner, more maintainable code (-18 lines).

joshuafuller commented:
Addressed the second round of feedback in commit acdd074:

**Resource Ownership:**
- Removed redundant `defer body.Close()` calls in streaming methods - `parseSSEStream` owns and closes the body

**Error Context:**
- Added contextual error wrapping in `unmarshalEvent` to indicate which event type failed parsing (Task, Message, TaskStatusUpdateEvent, TaskArtifactUpdateEvent)

**Test Error Handling:**
- Added explicit error checking to all JSON decode calls in tests (6 locations)

All tests continue to pass with race detection.

joshuafuller commented:
/gemini review

joshuafuller commented:
Addressed the third round of feedback in commit fdbe4ae:

**High Priority:**
- Fixed GetAgentCard to return copies of the cached card to prevent data races from concurrent modifications

**Medium Priority:**
- Refactored SendMessage to use unmarshalEvent helper (eliminates double unmarshaling for better efficiency)
- Extracted streamRequestToEvents helper method to eliminate code duplication between SendStreamingMessage and ResubscribeToTask

All tests pass with race detection. Net result: cleaner code with better efficiency (+31, -48 lines).

joshuafuller commented:
/gemini review

joshuafuller commented on a2aclient/jsonrpc.go:
```diff
@@ -0,0 +1,524 @@
+				}
+			}
+			// Ignore empty lines, comments, and other SSE event types
+		}
```
Implemented in e6f3fc1. Thanks for the suggestion!

yarolegovich commented on a2aclient/jsonrpc.go:
```diff
@@ -0,0 +1,524 @@
+}
+
+// parseSSEStream parses Server-Sent Events and yields JSON-RPC responses.
+func parseSSEStream(body io.ReadCloser) iter.Seq2[json.RawMessage, error] {
```
imo ownership transfer here introduces unnecessary coupling. I'd suggest to make `parseSSEStream` work with `io.Reader` and `defer body.Close()` in the caller instead

yarolegovich commented on a2aclient/jsonrpc.go:
```diff
@@ -0,0 +1,524 @@
+}
+
+// unmarshalEvent unmarshals a JSON-RPC result into an A2A Event.
+func unmarshalEvent(data json.RawMessage) (a2a.Event, error) {
```
fortunately this can be done in a cleaner way - in the spec there's a [`kind` field](https://github.com/a2aproject/A2A/blob/1a1e4de/types/src/types.ts#L785) which we can use. 
we omitted the field on runtime types to keep them cleaner. 

[this is how](https://github.com/a2aproject/a2a-go/blob/fee703e/a2a/core.go#L375) `kind` is attached back to `Part` types during marshaling. similar code is currently missing on `Message`, `Task` and `Task(Status|Artifact)UpdateEvent`, but we'll definitely need to add for jsonrpc server implementation.

[here is](https://github.com/a2aproject/a2a-go/blob/fee703e/a2a/core.go#L294) an example of how the field is used during unmarshaling.

given that this kind of decoding will be needed not just for jsonrpc client transport, I would implement it in a2a package, so something like:
```go
func UnmarshalEventJSON(json []byte) (Event, error) {...}
```
an alternative could be some kind of `EventJSON` or `RawEvent` struct, but a simple func is more obvious and easier to use

yarolegovich commented on a2aclient/jsonrpc.go:
```diff
@@ -0,0 +1,524 @@
+// to prevent data races from concurrent modifications.
+func (t *jsonrpcTransport) GetAgentCard(ctx context.Context) (*a2a.AgentCard, error) {
+	// Fast path: check if we have a card that doesn't need extension
+	t.cardMu.RLock()
```
I would probably not embed card caching on the transport layer:
1. it's is too low-level for making such decisions.
2. to provide consistent client behavior we would need to implement it for all transports. and since transports are pluggable we don't control all implementations.

I think it's better for consumers interested in this functionality to implement a cache on top of a2aclient where they'll have full control over which goroutine is trying to fetch, how the result is passed to other interested goroutine, how cache is invalidated etc.

joshuafuller replied:
Removed in 1061a48. Simplified GetAgentCard to just return the provided card without caching.

joshuafuller commented on a2aclient/jsonrpc.go:
```diff
@@ -0,0 +1,524 @@
+
+// streamRequestToEvents handles SSE streaming for JSON-RPC methods.
+// It converts the SSE stream into a sequence of A2A events.
+func (t *jsonrpcTransport) streamRequestToEvents(ctx context.Context, method string, params interface{}) iter.Seq2[a2a.Event, error] {
```
Done in 1061a48

joshuafuller commented:
Thanks for the detailed review! I've addressed all four comments in 1061a48:

  1. Replaced interface{} with any throughout
  2. Refactored parseSSEStream to accept io.Reader with caller-managed closing
  3. Implemented kind-based event discrimination following the existing ContentParts and TextPart patterns - added
  MarshalJSON methods to all Event types and created UnmarshalEventJSON() in the a2a package
  4. Removed agent card caching from the transport layer

  All tests pass with race detection enabled. I also added comprehensive test coverage for the new Event JSON
  functionality.

yarolegovich commented on a2a/event_json_test.go:
```diff
@@ -0,0 +1,373 @@
+		name      string
+		event     Event
+		wantKind  string
+		checkJSON func(t *testing.T, jsonStr string)
```
seems like checkJSON is repeating wantKind check everywhere. I'd probably just have `wantSubstrings []string` or something like that

joshuafuller commented on a2aclient/jsonrpc_test.go:
```diff
@@ -0,0 +1,755 @@
+	// due to the double-check pattern
+	if callCount > 3 {
+		t.Errorf("Expected at most 3 calls to extended card endpoint, got %d", callCount)
+	}
```
Fixed in fb75ea1 - changed assertion from `callCount > 3` to `callCount == 1` to properly verify the double-checked locking pattern ensures exactly one network call under concurrent load.

joshuafuller commented on a2aclient/jsonrpc.go:
```diff
@@ -0,0 +1,558 @@
+				// Function exiting normally - cleanup handled by defer
+			}
+		}()
+
```
Fixed in fb75ea1 - removed unnecessary context cancellation goroutines in `SendStreamingMessage` and `ResubscribeToTask`.

The `http.Client` respects context cancellation automatically when using `http.NewRequestWithContext`, and `Read` calls on `http.Response.Body` return an error if context is canceled (typically `context.Canceled`), which is caught by `scanner.Err()` in `parseSSEStream`.

Simplified code by ~12 lines.

joshuafuller commented on a2aclient/jsonrpc.go:
```diff
@@ -0,0 +1,541 @@
+			yield(nil, err)
+			return
+		}
+		defer body.Close()
```
Addressed in commit acdd074. This is the same issue as earlier comments about redundant defer body.Close(). Removed in both SendStreamingMessage and ResubscribeToTask.

joshuafuller commented on a2aclient/jsonrpc.go:
```diff
@@ -0,0 +1,541 @@
+			yield(nil, err)
+			return
+		}
+		defer body.Close()
```
Addressed in commit acdd074. Same resolution as earlier comments.

joshuafuller commented on a2aclient/jsonrpc.go:
```diff
@@ -0,0 +1,541 @@
+	}
+
+	return nil, fmt.Errorf("unknown event type: %s", string(data))
+}
```
Addressed in commit 1061a48. Replaced unmarshalEvent with a2a.UnmarshalEventJSON using kind-based discrimination per A2A spec.

joshuafuller commented on a2aclient/jsonrpc.go:
```diff
@@ -0,0 +1,541 @@
+	// Cache the extended card
+	t.agentCard = &extendedCard
+	return t.agentCard, nil
+}
```
Addressed in commit 1061a48

Removed agent card caching entirely from transport layer per maintainer feedback. Caching is too low-level for transport layer, and consumers interested in caching can implement it on top of a2aclient with full control over goroutine coordination, cache invalidation, etc.

Current implementation now simply returns the provided agent card without any caching or locking complexity.

joshuafuller commented on a2a/event_json_test.go:
```diff
@@ -0,0 +1,373 @@
+}
+
+// contains checks if a string contains a substring.
+func contains(s, substr string) bool {
```
Fixed in 504d08b - replaced custom contains() function with strings.Contains from standard library.

Changes:
- Removed custom contains() function (3 lines)
- Removed custom containsAt() helper (7 lines)
- Added import "strings"
- All substring checks now use stdlib strings.Contains()

Cleaner, more idiomatic Go code.

joshuafuller commented on a2aclient/jsonrpc.go:
```diff
@@ -0,0 +1,424 @@
+// WithJSONRPCTransport returns a Client factory option that enables JSON-RPC transport support.
+// When applied, the client will use JSON-RPC 2.0 over HTTP for all A2A protocol communication
+// as defined in the A2A specification §7.
+func WithJSONRPCTransport(opts ...JSONRPCOption) FactoryOption {
```
Fixed in 4ce661e - added WithJSONRPCTransport() to defaultOptions.

Transport ordering now matches other A2A SDKs:
- Python SDK: JSON-RPC as primary/fallback
- Java SDK: JSON-RPC included by default
- JavaScript SDK: jsonrpc_transport_handler.ts

Updated defaultOptions to: {WithJSONRPCTransport(), WithGRPCTransport()}
This ensures consistent behavior across all A2A client implementations.

joshuafuller commented on a2aclient/jsonrpc.go:
```diff
@@ -0,0 +1,424 @@
+			yield(nil, err)
+			return
+		}
+		defer func() { _ = body.Close() }()
```
Fixed in 504d08b - added TODO(yarolegovich) comments for all swallowed errors.

Added to 3 locations:
- Line 176: defer httpResp.Body.Close() in sendRequest
- Line 223: httpResp.Body.Close() in sendStreamingRequest error path
- Line 301: defer body.Close() in streamRequestToEvents

All swallowed errors now have TODO comments tracking future logging implementation.

joshuafuller commented on a2aclient/jsonrpc_test.go:
```diff
@@ -0,0 +1,712 @@
+		t.Fatal("Expected error, got nil")
+	}
+
+	jsonrpcErr, ok := err.(*jsonrpcError)
```
Added TODO in 626db1d for transport-agnostic error conversion.

Added TODO(yarolegovich) comment at jsonrpcError type definition to track the need for converting errors to transport-agnostic format. This will allow Client to use errors.Is(err, a2a.ErrMethodNotFound) patterns.

Leaving as TODO for now as you mentioned opening a ticket to address this across all transports.

yarolegovich commented on a2aclient/jsonrpc.go:
```diff
@@ -0,0 +1,524 @@
+
+	if httpResp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected HTTP status code: %d", httpResp.StatusCode)
+	}
```
we don't expect anything in the body, but I'd rather log status text than code:
```patch
--- a/a2aclient/jsonrpc.go
+++ b/a2aclient/jsonrpc.go
@@ -179,7 +179,7 @@ func (t *jsonrpcTransport) sendRequest(ctx context.Context, method string, param
        defer func() { _ = httpResp.Body.Close() }()

        if httpResp.StatusCode != http.StatusOK {
-               return nil, fmt.Errorf("unexpected HTTP status code: %d", httpResp.StatusCode)
+               return nil, fmt.Errorf("unexpected HTTP status code: %s", httpResp.Status)
        }
```
